unblob/pool.py

```diff
@@ -72,11 +83,30 @@ def __init__(
             for _ in range(process_num)
         ]
         self._tid = threading.get_native_id()
+        self._signals_fired = []
+
+    def _handle_signal(self, signum: int, frame):
+        logger.debug(f"Signal ({signal.Signals(signum).name}) received")
+        self._signals_fired.append((signum, frame))
+        raise ShutDownRequired
+
+    def _register_signal_handler(self, sig: signal.Signals):
+        return signal.signal(sig, self._handle_signal)
 
     def start(self):
         for p in self._procs:
             p.start()
 
+        # Save the original signal handlers while setting our own
```

Comment:
What if we add this signal handler setup-teardown machinery to its own context manager and just use it in the `process_until_done` call? That way we could separate its complexity from the pool's own complexity

Reply:
Separating this logic from the Pool's logic is a great idea, but as the part of "signal handler setup-teardown machinery", we are:
* registering our new signal handlers: this needs to be done as early as possible, but I can live with it if we want to do it before the `process_until_done`.
* calling the original signal handlers: sadly, we can not do to that right after the `process_until_done`, but after the last part of the `close` of the Pool, right now. 

This makes the whole context_manager over process_until_done concept not ideal, I think.

Reply:
Why need it to be in the close?

Reply:
Because if we (re-)raise the signal to the original handler, it will be handling it during the close, without allowing the Pool to gracefully exit.